### ISSUE TITLE
Connection Reset during Bulk API extract

### ISSUE THREAD

**Issue by bigassforce** [forcedotcom/dataloader#170]
Wide extracts with many fields cause hard failures by Connection Reset - here's a typical stack trace:

### [java] ExtractException: java.net.SocketException: Connection reset

```c#
[extract] progress.NihilistProgressAdapter setSubTask (NihilistProgressAdapter.java:78) Processed 58,000 of 35,444,913 total records. Rate: 804,000 records per hour. Estimated time to complete: 2,640 minutes and 22 seconds. There are 58,000 successes and 0 errors.
[java] 2018-11-20 00:33:37,604 ERROR [extract] action.AbstractAction handleException (AbstractAction.java:222) Exception occured during loading
[java] com.salesforce.dataloader.exception.ExtractException: java.net.SocketException: Connection reset
[java] at com.salesforce.dataloader.action.visitor.BulkQueryVisitor.writeExtractionBatch(BulkQueryVisitor.java:121)
[java] at com.salesforce.dataloader.action.visitor.BulkQueryVisitor.writeExtraction(BulkQueryVisitor.java:91)
[java] at com.salesforce.dataloader.action.visitor.AbstractQueryVisitor.visit(AbstractQueryVisitor.java:83)
[java] at com.salesforce.dataloader.action.AbstractExtractAction.visit(AbstractExtractAction.java:57)
[java] at com.salesforce.dataloader.action.AbstractAction.execute(AbstractAction.java:131)
[java] at com.salesforce.dataloader.controller.Controller.executeAction(Controller.java:156)
[java] at com.salesforce.dataloader.process.ProcessRunner.run(ProcessRunner.java:156)
[java] at com.salesforce.dataloader.process.ProcessRunner.run(ProcessRunner.java:107)
[java] at com.salesforce.dataloader.process.ProcessRunner.main(ProcessRunner.java:266)

[java] Caused by: java.net.SocketException: Connection reset
[java] at java.base/java.net.SocketInputStream.read(SocketInputStream.java:210)
[java] at java.base/java.net.SocketInputStream.read(SocketInputStream.java:141)
[java] at java.base/sun.security.ssl.SSLSocketInputRecord.read(SSLSocketInputRecord.java:425)
[java] at java.base/sun.security.ssl.SSLSocketInputRecord.bytesInCompletePacket(SSLSocketInputRecord.java:65)
[java] at java.base/sun.security.ssl.SSLSocketImpl.bytesInCompletePacket(SSLSocketImpl.java:918)
[java] at java.base/sun.security.ssl.AppInputStream.read(AppInputStream.java:144)
[java] at java.base/java.io.BufferedInputStream.fill(BufferedInputStream.java:252)
[java] at java.base/java.io.BufferedInputStream.read1(BufferedInputStream.java:292)
[java] at java.base/java.io.BufferedInputStream.read(BufferedInputStream.java:351)
[java] at java.base/sun.net.www.http.ChunkedInputStream.fastRead(ChunkedInputStream.java:244)
[java] at java.base/sun.net.www.http.ChunkedInputStream.read(ChunkedInputStream.java:689)
[java] at java.base/java.io.FilterInputStream.read(FilterInputStream.java:133)
[java] at java.base/sun.net.www.protocol.http.HttpURLConnection$HttpInputStream.read(HttpURLConnection.java:3444)
[java] at java.base/java.util.zip.InflaterInputStream.fill(InflaterInputStream.java:243)
[java] at java.base/java.util.zip.InflaterInputStream.read(InflaterInputStream.java:159)
[java] at java.base/java.util.zip.GZIPInputStream.read(GZIPInputStream.java:118)
[java] at java.base/sun.nio.cs.StreamDecoder.readBytes(StreamDecoder.java:284)
[java] at java.base/sun.nio.cs.StreamDecoder.implRead(StreamDecoder.java:326)
[java] at java.base/sun.nio.cs.StreamDecoder.read(StreamDecoder.java:178)
[java] at java.base/java.io.InputStreamReader.read(InputStreamReader.java:185)
[java] at java.base/java.io.BufferedReader.fill(BufferedReader.java:161)
[java] at java.base/java.io.BufferedReader.read(BufferedReader.java:182)
[java] at java.base/java.io.StreamTokenizer.read(StreamTokenizer.java:500)
[java] at java.base/java.io.StreamTokenizer.nextToken(StreamTokenizer.java:625)
[java] at com.sforce.async.CSVReader.nextRecordLocal(CSVReader.java:238)
[java] at com.sforce.async.CSVReader.nextRecord(CSVReader.java:166)
```

Connection is falling over on the HTTP stream while sending the data into a CSV reader. Not sure about the internal buffering, but in this case the main bottleneck is the tokenizing of the huge file.

Could the CSV reader be holding up the HTTP stream? Should we try writing the HTTP result chunk to a temporary CSV file first, so that SFDC can have its socket back sooner? See [BulkQueryVisitor.java](https://github.com/forcedotcom/dataloader/blob/cccf91d0953bc57c39f3d8d1eddeaeba942b746e/src/main/java/com/salesforce/dataloader/action/visitor/BulkQueryVisitor.java#L81-L112)

```java
@Override
protected void writeExtraction() throws AsyncApiException, ExtractException, DataAccessObjectException {
    if (this.batch.getState() == BatchStateEnum.Failed)
        throw new ExtractException("Batch failed: " + this.batch.getStateMessage());
    final QueryResultList results = getController().getBulkClient().getClient()
            .getQueryResultList(this.batch.getJobId(), this.batch.getId());

    for (final String resultId : results.getResult()) {
        if (getProgressMonitor().isCanceled()) return;
        try {
            final InputStream resultStream = getController().getBulkClient().getClient()
                    .getQueryResultStream(this.batch.getJobId(), this.batch.getId(), resultId);
```

**`              //download chunk as TempFile before CsvReader becomes bottleneck`**

```java
            try {
                final CSVReader rdr = new CSVReader(resultStream, Config.BULK_API_ENCODING);
                rdr.setMaxCharsInFile(Integer.MAX_VALUE);
                rdr.setMaxRowsInFile(Integer.MAX_VALUE);
                List<String> headers;
                headers = rdr.nextRecord();
                List<String> csvRow;
                while ((csvRow = rdr.nextRecord()) != null) {
                    final StringBuilder id = new StringBuilder();
                    final Row daoRow = getDaoRow(headers, csvRow, id);
                    addResultRow(daoRow, id.toString());
                }
            } finally {
                resultStream.close();
            }
        } catch (final IOException e) {
            throw new ExtractException(e);
        }
    }
}
```

**Comment by bigassforce** [comment]
# Working implementation

```java
try {
    final InputStream resultStream = getController().getBulkClient().getClient()
            .getQueryResultStream(this.batch.getJobId(), this.batch.getId(), resultId);
    
    //temp csv
    File tempFile = File.createTempFile("sdl", ".csv");
    String tempPath = tempFile.getAbsolutePath();
    getLogger().info("Downloading result chunk " + resultId + " to " + tempPath);
    
    //download result
    int bytesRead;
    byte[] buffer = new byte[8 * 1024];
    OutputStream writer = new FileOutputStream(tempFile);
    while ((bytesRead = resultStream.read(buffer)) != -1) {
        writer.write(buffer, 0, bytesRead);
    }
    
    //stream from csv file
    InputStream reader = new FileInputStream(new File(tempPath));
    
    try {
        //instead of from result
        final CSVReader rdr = new CSVReader(reader, Config.BULK_API_ENCODING);
        rdr.setMaxCharsInFile(Integer.MAX_VALUE);
        rdr.setMaxRowsInFile(Integer.MAX_VALUE);
        List<String> headers;
        headers = rdr.nextRecord();
        List<String> csvRow;
        while ((csvRow = rdr.nextRecord()) != null) {
            final StringBuilder id = new StringBuilder();
            final Row daoRow = getDaoRow(headers, csvRow, id);
            addResultRow(daoRow, id.toString());
        }
    } finally {
        //cleanup temp file
        resultStream.close();
        writer.close(); reader.close();
        Files.delete(Paths.get(tempPath));
    }
} catch (final IOException e) {
    throw new ExtractException(e);
}
```

**Comment by ashitsalesforce** [comment]
Thanks @bigassforce for having a working code that shows how to minimize the open connection time by saving results to a temporary file. Moving it to the code will require a few enhancements such as making sure there is sufficient space in the temp folder, generating a unique filename to avoid namespace collision in case same user or multiple users are running multiple copies of data loader, and creating related config params to enable/disable the feature. We are right now in a strict maintenance mode with only version upgrades and critical fixes. As such, I am leaving this issue open (assuming that open connection over long time is indeed the root-cause for the issue) for now.

**Comment by ashitsalesforce** [comment]
Closing this issue as the fix is introduced in PR #253 . Set loader.bufferUnprocessedBulkQueryResults to true in config.properties to buffer the query results.
